packages/agent/lib/agent.ts

```diff
@@ -0,0 +1,266 @@
+    super(opts)
+    // we will need this if they wish to make http requests over an https proxy
+    this.httpsAgent = new https.Agent({ keepAlive: true })
+  }
+
+  createSocket (req: http.ClientRequest, options: http.RequestOptions, cb: http.SocketCallback) {
+    if (process.env.HTTP_PROXY) {
+      const proxy = getProxyForUrl(options.href)
+
+      if (proxy) {
+        options.proxy = proxy
+
+        return this._createProxiedSocket(req, <RequestOptionsWithProxy>options, cb)
+      }
+    }
+
+    super.createSocket(req, options, cb)
+  }
+
+  _createProxiedSocket (req: http.ClientRequest, options: RequestOptionsWithProxy, cb: http.SocketCallback) {
+    debug(`Creating proxied socket for ${options.href} through ${options.proxy}`)
+
+    const proxy = url.parse(options.proxy)
+
+    // set req.path to the full path so the proxy can resolve it
+    // @ts-ignore: Cannot assign to 'path' because it is a constant or a read-only property.
+    req.path = options.href
+
+    delete req._header // so we can set headers again
+
+    req.setHeader('host', `${options.host}:${options.port}`)
+    if (proxy.auth) {
+      req.setHeader('proxy-authorization', `basic ${Buffer.from(proxy.auth).toString('base64')}`)
+    }
+
+    // node has queued an HTTP message to be sent already, so we need to regenerate the
+    // queued message with the new path and headers
+    // https://github.com/TooTallNate/node-http-proxy-agent/blob/master/index.js#L93
+    regenerateRequestHead(req)
+
+    options.port = Number(proxy.port || 80)
+    options.host = proxy.hostname || 'localhost'
+    delete options.path // so the underlying net.connect doesn't default to IPC
+
+    if (proxy.protocol === 'https:') {
+      // gonna have to use the https module to reach the proxy, even though this is an http req
+      req.agent = this.httpsAgent
+
+      return this.httpsAgent.addRequest(req, options)
+    }
+
+    super.createSocket(req, options, cb)
+  }
+}
+
+class HttpsAgent extends https.Agent {
+  constructor (opts: https.AgentOptions = {}) {
+    opts.keepAlive = true
+    super(opts)
+  }
+
+  createConnection (options: http.RequestOptions, cb: http.SocketCallback) {
+    if (process.env.HTTPS_PROXY) {
+      const proxy = getProxyForUrl(options.href)
+
+      if (typeof proxy === "string") {
+        options.proxy = <string>proxy
+
+        return this.createProxiedConnection(<RequestOptionsWithProxy>options, cb)
+      }
+    }
+
+    // @ts-ignore
+    cb(null, super.createConnection(options))
+  }
+
+  createProxiedConnection (options: RequestOptionsWithProxy, cb: http.SocketCallback) {
+    // heavily inspired by
+    // https://github.com/mknj/node-keepalive-proxy-agent/blob/master/index.js
+    debug(`Creating proxied socket for ${options.href} through ${options.proxy}`)
+
+    const proxy = url.parse(options.proxy)
+    const port = options.uri.port || 443
+    const hostname = options.uri.hostname
+
+    const proxySocket = createProxySock(proxy)
```

Comment:
why doesn't this have the `setNoDelay` option set?

Reply:
It doesn't seem to need it, the performance doesn't change one way or another (tested using the performance tests from earlier).